lib/service/db.go

```diff
@@ -0,0 +1,212 @@
+			}))
+	}
+
+	authorizer, err := auth.NewAuthorizer(conn.Client, conn.Client, conn.Client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tlsConfig, err := conn.ServerIdentity.TLSConfig(process.Config.CipherSuites)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	asyncEmitter, err := process.newAsyncEmitter(conn.Client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if retErr != nil {
+			warnOnErr(asyncEmitter.Close(), process.log)
+		}
+	}()
+
+	streamer, err := events.NewCheckingStreamer(events.CheckingStreamerConfig{
+		Inner: conn.Client,
+		Clock: process.Clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create and start the database service.
+	dbService, err := db.New(process.ExitContext(), db.Config{
+		DataDir:     process.Config.DataDir,
+		AuthClient:  conn.Client,
+		AccessPoint: accessPoint,
+		StreamEmitter: &events.StreamerAndEmitter{
+			Emitter:  asyncEmitter,
+			Streamer: streamer,
+		},
+		Authorizer:  authorizer,
+		TLSConfig:   tlsConfig,
+		GetRotation: process.getRotation,
+		Servers:     databaseServers,
+		OnHeartbeat: func(err error) {
+			if err != nil {
+				process.BroadcastEvent(Event{Name: TeleportDegradedEvent, Payload: teleport.ComponentDatabase})
+			} else {
+				process.BroadcastEvent(Event{Name: TeleportOKEvent, Payload: teleport.ComponentDatabase})
+			}
+		},
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := dbService.Start(); err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if retErr != nil {
+			warnOnErr(dbService.Close(), process.log)
+		}
+	}()
+
+	// Create and start the agent pool.
+	agentPool, err := reversetunnel.NewAgentPool(process.ExitContext(),
+		reversetunnel.AgentPoolConfig{
+			Component:   teleport.ComponentDatabase,
+			HostUUID:    conn.ServerIdentity.ID.HostUUID,
+			ProxyAddr:   tunnelAddr,
+			Client:      conn.Client,
+			Server:      dbService,
+			AccessPoint: conn.Client,
+			HostSigner:  conn.ServerIdentity.KeySigner,
+			Cluster:     conn.ServerIdentity.Cert.Extensions[utils.CertExtensionAuthority],
+		})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := agentPool.Start(); err != nil {
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if retErr != nil {
+			agentPool.Stop()
+		}
+	}()
+
+	process.BroadcastEvent(Event{Name: DatabasesReady, Payload: nil})
+	log.Infof("Database service has successfully started: %v.", databaseServers)
+
+	// Block and wait while the server and agent pool are running.
+	if err := dbService.Wait(); err != nil {
+		return trace.Wrap(err)
+	}
+	agentPool.Wait()
+
+	// Execute this when the process running database proxy service exits.
+	process.onExit("db.stop", func(payload interface{}) {
+		log.Info("Shutting down.")
+		if dbService != nil {
+			warnOnErr(dbService.Close(), process.log)
+		}
+		if asyncEmitter != nil {
+			warnOnErr(asyncEmitter.Close(), process.log)
+		}
+		if agentPool != nil {
+			agentPool.Stop()
+		}
+		log.Info("Exited.")
+	})
```

Comment:
should this callback be registered before `dbService.Wait`?

Reply:
Good call.

Reply:
@awly Fixed.